sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -1014,7 +1016,9 @@ object TypeCoercion extends TypeCoercionBase {
       case (_: StringType, datetime: DatetimeType) => datetime
       case (_: StringType, AnyTimestampType) => AnyTimestampType.defaultConcreteType
       case (_: StringType, BinaryType) => BinaryType
-      // Cast any atomic type to string.
+      // Cast any atomic type to string except if there are strings with different collations. In
+      // that case we skip the cast.
+      case (st1: StringType, st2: StringType) if st1.collationId != st2.collationId => null
```

Comment:
I think this makes sense. Just for my curiosity, can we really hit this branch today? I don't think there is any function that only accept a certion string collation.

Reply:
I updated a comment, agree that we should not hit this branch in any execution

Reply:
This seems like we are adding something unnecessary. The comment said we cast any atomic type and that rule is checking for whether any is instance of StringType, so we cannot hit this by any rule below actually, so we should be fine with getting the default which is null? I am good with adding additional comment as to why we do not cast StringType in the existing rule, but this is just moving the default branch earlier.